File: vertical/3.3.0/csharp/table/Adapter.cs
// Reconstructed demonstration binding, not a historical production release.
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Encodings.Web;

static class Adapter {
    static readonly JsonSerializerOptions Options = new() { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
    static readonly string[][] Roots = {
        new[]{"tableId","tableName","title","stringValues","numericValues"},
        new[]{"columns","rows","styleMatrix"},
        new[]{"sparseOverride","sparseOverrideTypes"},
        new[]{"stylebook","fontConfig"},
        new[]{"imageManager","cellBackgroundImageRefs","cellIds","dataBarScale","cellDataBars","captureTokens"},
        new[]{"columnPriority","rowPriority","hiddenColumns","hiddenRows"}
    };
    static readonly string[] Forbidden = {"allStylesLight","allStylesDark","allStylesExtra","themes"};
    static void Require(bool ok, string code) { if (!ok) throw new InvalidDataException(code); }
    static bool Str(JsonNode? n) => n is JsonValue v && v.TryGetValue<string>(out _);
    static bool Num(JsonNode? n) => n is JsonValue v && v.TryGetValue<double>(out var d) && double.IsFinite(d);
    static int Integer(JsonNode? n, string code) {
        Require(Num(n), code); double d=n!.GetValue<double>();
        Require(d==Math.Truncate(d) && d>=int.MinValue && d<=int.MaxValue, code); return (int)d;
    }
    static string Pointer(string key) => key.Replace("~","~0").Replace("/","~1");
    static void Unique(JsonElement e) {
        if(e.ValueKind==JsonValueKind.Object) {
            var keys=new HashSet<string>();
            foreach(var p in e.EnumerateObject()) { Require(keys.Add(p.Name),"json"); Unique(p.Value); }
        } else if(e.ValueKind==JsonValueKind.Array) foreach(var x in e.EnumerateArray()) Unique(x);
    }
    static void Project(JsonObject obj, HashSet<string> fields, bool extra, string path, List<string> lost) {
        foreach(var key in obj.Select(p=>p.Key).ToArray()) {
            if((!fields.Contains(key) && !extra) || (fields.Contains(key) && obj[key] is null)) {
                obj.Remove(key); lost.Add(path+"/"+Pointer(key));
            }
        }
    }
    static void Metadata(JsonObject table,string name,bool column,List<string> lost) {
        if(table[name] is not JsonNode node) return;
        Require(node is JsonArray,"metadata");
        var fields=new HashSet<string>(column ? new[]{"id","header","typeHint","width","alignment"} : new[]{"id","header","height"});
        if(Profile.Level==5) { fields.Add(column?"widthExpr":"heightExpr"); fields.Add("hidden"); }
        int i=0;
        foreach(var item in (JsonArray)node) {
            Require(item is JsonObject,"metadata"); var row=(JsonObject)item!;
            Project(row,fields,Profile.Level==5,$"/{name}/{i++}",lost);
            foreach(var key in fields) if(row[key] is JsonNode value) {
                if(key=="hidden") Require(value is JsonValue v && v.TryGetValue<bool>(out _),"metadata");
                else if(key.EndsWith("Expr")) Require(value is JsonObject,"expression");
                else Require(Str(value),"metadata");
            }
        }
    }
    public static JsonObject Process(string line) {
        try {
            using var doc=JsonDocument.Parse(line,new JsonDocumentOptions{MaxDepth=64}); Unique(doc.RootElement);
            var node=JsonNode.Parse(line); Require(node is JsonObject,"table"); var table=(JsonObject)node!;
            var lost=new List<string>();
            if(Profile.Level>=3) Require(!Forbidden.Any(table.ContainsKey),"forbidden-key");
            var fields=Roots.Take(Profile.Level+1).SelectMany(x=>x).ToHashSet();
            Project(table,fields,Profile.Level>=3,"",lost);
            foreach(var key in new[]{"tableId","tableName","title"}) if(table[key] is JsonNode n) Require(Str(n),"identity");
            int? rows=null,cols=null;
            foreach(var key in new[]{"stringValues","numericValues","styleMatrix"}.Where(fields.Contains)) {
                if(table[key] is not JsonNode grid) continue; Require(grid is JsonArray,"grid-shape");
                int r=((JsonArray)grid).Count,c=-1;
                foreach(var row in (JsonArray)grid) {
                    Require(row is JsonArray,"grid-shape"); var a=(JsonArray)row!;
                    if(c<0)c=a.Count; Require(a.Count==c,"grid-shape");
                    foreach(var value in a) Require(value is null || (key=="numericValues" ? Num(value) : Str(value)),"cell-type");
                }
                c=Math.Max(0,c); Require(rows is null || (rows==r && cols==c),"grid-shape"); rows=r;cols=c;
            }
            if(Profile.Level>=1) { Metadata(table,"columns",true,lost);Metadata(table,"rows",false,lost); }
            if(Profile.Level>=2 && table["sparseOverride"] is JsonNode sparse) {
                Require(sparse is JsonObject,"sparse"); var occupied=new HashSet<(int,int)>();
                var cellFields=new HashSet<string>(new[]{"rowIndex","colIndex","rowSpan","colSpan","value","valueNumeric","styleKey","typeRef","formula","formatString"});
                if(Profile.Level>=4) cellFields.UnionWith(new[]{"imageRef","imageAlt","imageFit","imageZoomable","videoRef","posterRef","videoControls","videoMuted","videoLoop","videoFit","mathTex","mathAst"});
                foreach(var entry in (JsonObject)sparse) {
                    Require(entry.Value is JsonArray,"sparse"); int i=0;
                    foreach(var item in (JsonArray)entry.Value!) {
                        Require(item is JsonObject,"sparse"); var cell=(JsonObject)item!;
                        Project(cell,cellFields,Profile.Level==5,"/sparseOverride/"+Pointer(entry.Key)+"/"+i++,lost);
                        int r=Integer(cell["rowIndex"],"coordinate"),c=Integer(cell["colIndex"],"coordinate");
                        Require(r>=0 && c>=0 && (rows is null || (r<rows && c<cols)),"coordinate");
                        int rs=cell["rowSpan"] is null?1:Integer(cell["rowSpan"],"span"),cs=cell["colSpan"] is null?1:Integer(cell["colSpan"],"span");
                        Require(rs>0 && cs>0 && (long)rs*cs<=100000 && (rows is null || ((long)r+rs<=rows && (long)c+cs<=cols)),"span");
                        if(rs>1 || cs>1) for(int y=0;y<rs;y++) for(int x=0;x<cs;x++) Require(occupied.Add((r+y,c+x)),"merge-overlap");
                        foreach(var p in cell.Where(p=>cellFields.Contains(p.Key))) {
                            if(new[]{"valueNumeric"}.Contains(p.Key)) Require(Num(p.Value),"cell-type");
                            else if(new[]{"imageZoomable","videoControls","videoMuted","videoLoop"}.Contains(p.Key)) Require(p.Value is JsonValue v && v.TryGetValue<bool>(out _),"cell-type");
                            else if(p.Key=="mathAst") Require(p.Value is JsonObject,"expression");
                            else if(!new[]{"rowIndex","colIndex","rowSpan","colSpan"}.Contains(p.Key)) Require(Str(p.Value),"cell-type");
                        }
                    }
                }
            }
            if(Profile.Level>=2 && table["sparseOverrideTypes"] is JsonNode types) {
                Require(types is JsonObject,"type-hint");
                foreach(var p in (JsonObject)types) Require(Str(p.Value) && new[]{"text","number","date","boolean","formula","image"}.Contains(p.Value!.GetValue<string>()),"type-hint");
            }
            if(Profile.Level>=3 && table["stylebook"] is JsonNode book) {
                Require(book is JsonObject,"style"); foreach(var p in (JsonObject)book) {
                    Require(p.Value is JsonObject,"style"); var style=(JsonObject)p.Value!;
                    if(style["weight"] is JsonNode w) { int n=Integer(w,"style");Require(n>=0 && n<=1000,"style"); }
                    if(style["align"] is JsonNode a) Require(Str(a) && new[]{"left","right","center","justify","decimal"}.Contains(a.GetValue<string>()),"style");
                    if(style["numFormat"] is JsonNode f) Require(Str(f),"style");
                }
            }
            if(Profile.Level>=4 && table["dataBarScale"] is JsonNode scale) {
                Require(scale is JsonObject,"scale"); var s=(JsonObject)scale;
                foreach(var k in new[]{"min","max","zero"}) if(s[k] is JsonNode v) Require(Num(v),"scale");
                double min=s["min"]?.GetValue<double>()??-100,max=s["max"]?.GetValue<double>()??100,zero=s["zero"]?.GetValue<double>()??0;
                Require(min<max && zero>=min && zero<=max,"scale");
            }
            if(Profile.Level==5) foreach(var key in new[]{"columnPriority","rowPriority","hiddenColumns","hiddenRows"}) {
                if(table[key] is not JsonNode arr) continue; Require(arr is JsonArray,"visibility");
                var seen=new HashSet<int>(); foreach(var v in (JsonArray)arr) { int n=Integer(v,"visibility");
                    if(key.StartsWith("hidden")) Require(n>=0 && seen.Add(n) && (rows is null || n<(key=="hiddenColumns"?cols:rows)),"visibility");
                }
            }
            lost.Sort(StringComparer.Ordinal);
            return new JsonObject{{"ok",true},{"table",table},{"lost",JsonSerializer.SerializeToNode(lost)}};
        } catch(InvalidDataException e) { return new JsonObject{{"ok",false},{"error",e.Message}}; }
        catch(JsonException) { return new JsonObject{{"ok",false},{"error","json"}}; }
    }
    static void Main() {
        Console.InputEncoding=new System.Text.UTF8Encoding(false);
        Console.OutputEncoding=new System.Text.UTF8Encoding(false);
        string? line;while((line=Console.ReadLine())!=null) Console.WriteLine(Process(line).ToJsonString(Options));
    }
}

File: evolution/templates/adapter.rs
// Reconstructed demonstration binding, not a historical production release.
use serde::de::{self, Deserialize, Deserializer, MapAccess, SeqAccess, Visitor};
use serde_json::{json, Map, Value};
use std::{collections::BTreeSet, fmt, io::{self, BufRead}};
mod profile;
use profile::LEVEL;
type Result<T> = std::result::Result<T, &'static str>;
const ROOTS: &[&[&str]] = &[
    &["tableId","tableName","title","stringValues","numericValues"],
    &["columns","rows","styleMatrix"],
    &["sparseOverride","sparseOverrideTypes"],
    &["stylebook","fontConfig"],
    &["imageManager","cellBackgroundImageRefs","cellIds","dataBarScale","cellDataBars","captureTokens"],
    &["columnPriority","rowPriority","hiddenColumns","hiddenRows"],
];
fn require(ok:bool, code:&'static str)->Result<()> {if ok {Ok(())} else {Err(code)}}
fn integer(v:Option<&Value>, code:&'static str)->Result<i32> {
    let d=v.and_then(Value::as_f64).ok_or(code)?;
    require(d.is_finite() && d.fract()==0.0 && d>=i32::MIN as f64 && d<=i32::MAX as f64,code)?; Ok(d as i32)
}
fn pointer(s:&str)->String {s.replace('~',"~0").replace('/',"~1")}
fn project(obj:&mut Map<String,Value>,fields:&BTreeSet<&str>,extra:bool,path:&str,lost:&mut Vec<String>) {
    let keys:Vec<_>=obj.iter().filter(|(k,v)|(!fields.contains(k.as_str()) && !extra)||(fields.contains(k.as_str()) && v.is_null())).map(|(k,_)|k.clone()).collect();
    for k in keys {obj.remove(&k);lost.push(format!("{path}/{}",pointer(&k)));}
}
fn metadata(table:&mut Map<String,Value>,name:&str,column:bool,lost:&mut Vec<String>)->Result<()> {
    if let Some(node)=table.get_mut(name) {
        let rows=node.as_array_mut().ok_or("metadata")?;
        let mut fields:BTreeSet<&str>=if column {vec!["id","header","typeHint","width","alignment"]}else{vec!["id","header","height"]}.into_iter().collect();
        if LEVEL==5 {fields.insert(if column {"widthExpr"}else{"heightExpr"});fields.insert("hidden");}
        for (i,item) in rows.iter_mut().enumerate() {
            let row=item.as_object_mut().ok_or("metadata")?;project(row,&fields,LEVEL==5,&format!("/{name}/{i}"),lost);
            for k in &fields {if let Some(v)=row.get(*k) {
                if *k=="hidden" {require(v.is_boolean(),"metadata")?;}
                else if k.ends_with("Expr") {require(v.is_object(),"expression")?;}
                else {require(v.is_string(),"metadata")?;}
            }}
        }
    } Ok(())
}
fn process(mut node:Value)->Result<Value> {
    let table=node.as_object_mut().ok_or("table")?;let mut lost=vec![];
    if LEVEL>=3 {require(!["allStylesLight","allStylesDark","allStylesExtra","themes"].iter().any(|k|table.contains_key(*k)),"forbidden-key")?;}
    let fields:BTreeSet<&str>=ROOTS.iter().take(LEVEL+1).flat_map(|x|x.iter().copied()).collect();
    project(table,&fields,LEVEL>=3,"",&mut lost);
    for k in ["tableId","tableName","title"] {if let Some(v)=table.get(k) {require(v.is_string(),"identity")?;}}
    let mut shape=None;
    for k in ["stringValues","numericValues","styleMatrix"] {if fields.contains(k) {if let Some(grid)=table.get(k) {
        let rows=grid.as_array().ok_or("grid-shape")?;let mut cols=None;
        for row in rows {let cells=row.as_array().ok_or("grid-shape")?;
            require(cols.is_none() || cols==Some(cells.len()),"grid-shape")?;cols=Some(cells.len());
            for v in cells {require(v.is_null() || if k=="numericValues" {v.as_f64().is_some_and(f64::is_finite)}else{v.is_string()},"cell-type")?;}
        }
        let now=(rows.len(),cols.unwrap_or(0));require(shape.is_none() || shape==Some(now),"grid-shape")?;shape=Some(now);
    }}}
    if LEVEL>=1 {metadata(table,"columns",true,&mut lost)?;metadata(table,"rows",false,&mut lost)?;}
    if LEVEL>=2 {if let Some(sparse)=table.get_mut("sparseOverride") {
        let groups=sparse.as_object_mut().ok_or("sparse")?;let mut occupied=BTreeSet::new();
        let mut cell_fields:BTreeSet<&str>=["rowIndex","colIndex","rowSpan","colSpan","value","valueNumeric","styleKey","typeRef","formula","formatString"].into_iter().collect();
        if LEVEL>=4 {cell_fields.extend(["imageRef","imageAlt","imageFit","imageZoomable","videoRef","posterRef","videoControls","videoMuted","videoLoop","videoFit","mathTex","mathAst"]);}
        for (key,items) in groups {for(i,item)in items.as_array_mut().ok_or("sparse")?.iter_mut().enumerate() {
            let cell=item.as_object_mut().ok_or("sparse")?;project(cell,&cell_fields,LEVEL==5,&format!("/sparseOverride/{}/{i}",pointer(key)),&mut lost);
            let r=integer(cell.get("rowIndex"),"coordinate")?;let c=integer(cell.get("colIndex"),"coordinate")?;
            require(r>=0 && c>=0 && shape.is_none_or(|(rows,cols)| (r as usize)<rows && (c as usize)<cols),"coordinate")?;
            let rs=if cell.contains_key("rowSpan"){integer(cell.get("rowSpan"),"span")?}else{1};
            let cs=if cell.contains_key("colSpan"){integer(cell.get("colSpan"),"span")?}else{1};
            require(rs>0 && cs>0 && (rs as i64)*(cs as i64)<=100000 && shape.is_none_or(|(rows,cols)|r as i64+rs as i64<=rows as i64 && c as i64+cs as i64<=cols as i64),"span")?;
            if rs>1 || cs>1 {for y in 0..rs {for x in 0..cs {require(occupied.insert((r as i64+y as i64,c as i64+x as i64)),"merge-overlap")?;}}}
            for (k,v) in cell.iter().filter(|(k,_)|cell_fields.contains(k.as_str())) {
                match k.as_str() {
                    "valueNumeric"=>require(v.as_f64().is_some_and(f64::is_finite),"cell-type")?,
                    "imageZoomable"|"videoControls"|"videoMuted"|"videoLoop"=>require(v.is_boolean(),"cell-type")?,
                    "mathAst"=>require(v.is_object(),"expression")?,
                    "rowIndex"|"colIndex"|"rowSpan"|"colSpan"=>{},
                    _=>require(v.is_string(),"cell-type")?,
                }
            }
        }}
    }}
    if LEVEL>=2 {if let Some(types)=table.get("sparseOverrideTypes") {
        for (_,v) in types.as_object().ok_or("type-hint")? {require(v.as_str().is_some_and(|s|["text","number","date","boolean","formula","image"].contains(&s)),"type-hint")?;}
    }}
    if LEVEL>=3 {if let Some(book)=table.get("stylebook") {
        for(_,v)in book.as_object().ok_or("style")? {
            let s=v.as_object().ok_or("style")?;
            if s.contains_key("weight") {let n=integer(s.get("weight"),"style")?;require((0..=1000).contains(&n),"style")?;}
            if let Some(a)=s.get("align") {require(a.as_str().is_some_and(|s|["left","right","center","justify","decimal"].contains(&s)),"style")?;}
            if let Some(f)=s.get("numFormat") {require(f.is_string(),"style")?;}
        }
    }}
    if LEVEL>=4 {if let Some(scale)=table.get("dataBarScale") {
        let s=scale.as_object().ok_or("scale")?;
        for k in ["min","max","zero"] {if let Some(v)=s.get(k) {require(v.as_f64().is_some_and(f64::is_finite),"scale")?;}}
        let min=s.get("min").and_then(Value::as_f64).unwrap_or(-100.0);let max=s.get("max").and_then(Value::as_f64).unwrap_or(100.0);let zero=s.get("zero").and_then(Value::as_f64).unwrap_or(0.0);
        require(min<max && zero>=min && zero<=max,"scale")?;
    }}
    if LEVEL==5 {for k in ["columnPriority","rowPriority","hiddenColumns","hiddenRows"] {if let Some(arr)=table.get(k) {
        let mut seen=BTreeSet::new();for v in arr.as_array().ok_or("visibility")? {let n=integer(Some(v),"visibility")?;
            if k.starts_with("hidden") {require(n>=0 && seen.insert(n) && shape.is_none_or(|(rows,cols)|(n as usize)<if k=="hiddenColumns"{cols}else{rows}),"visibility")?;}
        }
    }}}
    lost.sort();Ok(json!({"ok":true,"table":node,"lost":lost}))
}
// Strict recursive JSON reader: reject duplicates instead of silently taking the last value.
struct Strict(Value);
impl<'de> Deserialize<'de> for Strict {
    fn deserialize<D:Deserializer<'de>>(d:D)->std::result::Result<Self,D::Error> {
        struct V;
        impl<'de> Visitor<'de> for V {
            type Value=Strict;
            fn expecting(&self,f:&mut fmt::Formatter)->fmt::Result {write!(f,"JSON")}
            fn visit_bool<E:de::Error>(self,v:bool)->std::result::Result<Strict,E>{Ok(Strict(json!(v)))}
            fn visit_i64<E:de::Error>(self,v:i64)->std::result::Result<Strict,E>{Ok(Strict(json!(v)))}
            fn visit_u64<E:de::Error>(self,v:u64)->std::result::Result<Strict,E>{Ok(Strict(json!(v)))}
            fn visit_f64<E:de::Error>(self,v:f64)->std::result::Result<Strict,E>{if !v.is_finite(){return Err(E::custom("nonfinite"));}Ok(Strict(json!(v)))}
            fn visit_str<E:de::Error>(self,v:&str)->std::result::Result<Strict,E>{Ok(Strict(json!(v)))}
            fn visit_unit<E:de::Error>(self)->std::result::Result<Strict,E>{Ok(Strict(Value::Null))}
            fn visit_none<E:de::Error>(self)->std::result::Result<Strict,E>{Ok(Strict(Value::Null))}
            fn visit_seq<A:SeqAccess<'de>>(self,mut a:A)->std::result::Result<Strict,A::Error>{let mut v=vec![];while let Some(Strict(x))=a.next_element()?{v.push(x);}Ok(Strict(Value::Array(v)))}
            fn visit_map<A:MapAccess<'de>>(self,mut a:A)->std::result::Result<Strict,A::Error>{let mut m=Map::new();while let Some((k,Strict(v)))=a.next_entry::<String,Strict>()?{if m.insert(k,v).is_some(){return Err(de::Error::custom("duplicate"));}}Ok(Strict(Value::Object(m)))}
        }
        d.deserialize_any(V)
    }
}
fn main() {
    for line in io::stdin().lock().lines() {let line=line.expect("stdin");
        let result=serde_json::from_str::<Strict>(&line).map_err(|_|"json").and_then(|Strict(v)|process(v));
        let value=match result {Ok(v)=>v,Err(e)=>json!({"ok":false,"error":e})};println!("{value}");
    }
}
